### PR TITLE
Support baremetal

### PR DESCRIPTION
**Pull request by ghost**
Add baremetal service functionalities.

### CODE REVIEW

File: ecl/baremetal/v2/servers/doc.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	allServers, err := servers.ExtractServers(allPages)
+	if err != nil {
+		panic(err)
+	}
+
+	for _, server := range allServers {
+		fmt.Printf("%+v", server)
+	}
+*/
+package servers
```
Comment (keiichi-hikita) [inline ecl/baremetal/v2/servers/doc.go:120]: Please write examples to delete servers.

File: ecl/baremetal/v2/availabilityzones/doc.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/*
+Package availabilityzones provides the ability to get lists and detailed
+availability zone information and to extend a server result with
+availability zone information.
+
+Example of Get Availability Zone Information
+
+	allPages, err := availabilityzones.List(computeClient).AllPages()
```
Comment (t-wata) [inline ecl/baremetal/v2/availabilityzones/doc.go]: `baremetalClient` ?

File: ecl/baremetal/v2/availabilityzones/testing/fixtures.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+}
+`
+
+var azResult = []az.AvailabilityZone{
+	{
+		Hosts:     nil,
+		ZoneName:  "zone1-groupa",
+		ZoneState: az.ZoneState{Available: true},
+	},
+}
```
Comment (t-wata) [inline ecl/baremetal/v2/availabilityzones/testing/fixtures.go]: Why check only "zone1-groupa" even though  "zone1-groupb" is also included in  var "actual" ?

File: ecl/baremetal/v2/flavors/requests.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+// ListOptsBuilder allows extensions to add additional parameters to the List
+// request.
+type ListOptsBuilder interface {
+	ToFlavorListQuery() (string, error)
+}
+
+// ListOpts holds options for listing flavors.
+// It is passed to the flavors.List function.
+type ListOpts struct {
+	// Now there are no definiton as query params in API specification
```
Comment (t-wata) [inline ecl/baremetal/v2/flavors/requests.go]: Typo: definition

File: ecl/baremetal/v2/flavors/requests.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		}
+		url += query
+	}
+
+	return pagination.NewPager(client, url, func(r pagination.PageResult) pagination.Page {
+		return FlavorPage{pagination.LinkedPageBase{PageResult: r}}
+	})
+}
+
+// IDFromName is a convienience function that returns a flavor's ID given its
```
Comment (t-wata) [inline ecl/baremetal/v2/flavors/requests.go]: Typo: convenience

File: ecl/baremetal/v2/flavors/requests.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+// It is passed to the flavors.List function.
+type ListOpts struct {
+	// Now there are no definiton as query params in API specification
+	// But do not remove this struct in future specification change.
+}
+
+// ToFlavorListQuery formats a ListOpts into a query string.
+func (opts ListOpts) ToFlavorListQuery() (string, error) {
+	q, err := eclcloud.BuildQueryString(opts)
+	return q.String(), err
```
Comment (t-wata) [inline ecl/baremetal/v2/flavors/requests.go:32]: I think it is necessary to check if "q" is nil.

File: ecl/client.go [hunk 7]
```diff
@@ -235,6 +235,12 @@ func NewComputeV2(client *eclcloud.ProviderClient, eo eclcloud.EndpointOpts) (*e
 	return initClientOpts(client, eo, "compute")
 }
 
+// NewBaremetalV2 creates a ServiceClient that may be used with the v2 compute
```
Comment (t-wata) [inline ecl/client.go]: `v2 compute` → `v2 baremetal`

File: ecl/baremetal/v2/servers/requests.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+	Marker string `q:"marker"`
+
+	// Limit is an integer value for the limit of values to return.
+	Limit int `q:"limit"`
+}
+
+// ToServerListQuery formats a ListOpts into a query string.
+func (opts ListOpts) ToServerListQuery() (string, error) {
+	q, err := eclcloud.BuildQueryString(opts)
+	return q.String(), err
```
Comment (t-wata) [inline ecl/baremetal/v2/servers/requests.go:53]: I think it is necessary to check if "q" is nil.

File: ecl/baremetal/v2/servers/requests.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+import (
+	"encoding/base64"
+
+	"github.com/nttcom/eclcloud"
+	"github.com/nttcom/eclcloud/pagination"
+)
+
+// Get retrieves the flavor with the provided ID.
+// To extract the Flavor object from the response,
```
Comment (t-wata) [inline ecl/baremetal/v2/servers/requests.go]: `flavor` -> `server`

File: ecl/baremetal/v2/servers/requests.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+// ListOptsBuilder allows extensions to add additional parameters to the List
+// request.
+type ListOptsBuilder interface {
+	ToServerListQuery() (string, error)
+}
+
+// ListOpts holds options for listing flavors.
+// It is passed to the flavors.List function.
```
Comment (t-wata) [inline ecl/baremetal/v2/servers/requests.go]: `flavor` -> `server`

File: ecl/baremetal/v2/servers/requests.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+	Limit int `q:"limit"`
+}
+
+// ToServerListQuery formats a ListOpts into a query string.
+func (opts ListOpts) ToServerListQuery() (string, error) {
+	q, err := eclcloud.BuildQueryString(opts)
+	return q.String(), err
+}
+
+// List returns Flavor optionally limited by the conditions provided in ListOpts.
```
Comment (t-wata) [inline ecl/baremetal/v2/servers/requests.go]: `Flavor` -> `Server`

File: ecl/baremetal/v2/servers/result.go [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+type Partition struct {
+	LVM            bool   `json:"lvm"`
+	Size           int    `json:"size"`
+	PartitionLabel string `json:"partition_label"`
+}
+
+// LVMVolumeGroup represents LVM volume group configuration for the server resource.
+type LVMVolumeGroup struct {
+	VGLabel                       string          `json:"vg_label"`
+	PhysicalVolumePartisionLabels []string        `json:"physical_volume_partition_labels"`
```
Comment (t-wata) [inline ecl/baremetal/v2/servers/result.go]: Typo: `PhysicalVolumePartitionLabels`

File: ecl/baremetal/v2/servers/requests.go [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+type CreateOptsPartition struct {
+	LVM            bool   `json:"lvm,omitempty"`
+	Size           string `json:"size,omitempty"`
+	PartitionLabel string `json:"partition_label,omitempty"`
+}
+
+// CreateOptsLVMVolumeGroup represents LVM volume group configuration for the server resource.
+type CreateOptsLVMVolumeGroup struct {
+	VGLabel                       string                    `json:"vg_label,omitempty"`
+	PhysicalVolumePartisionLabels []string                  `json:"physical_volume_partition_labels,omitempty"`
```
Comment (t-wata) [inline ecl/baremetal/v2/servers/requests.go]: Typo: `PhysicalVolumePartitionLabels`

File: ecl/baremetal/v2/servers/testing/requests_test.go [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+				UUID:    "d32019d3-bc6e-4319-9c1d-6722fc136a22",
+				FixedIP: "10.0.0.100",
+			},
+		},
+		AdminPass:        "aabbccddeeff",
+		ImageRef:         "b5660a6e-4b46-4be3-9707-6b47221b454f",
+		FlavorRef:        "05184ba3-00ba-4fbc-b7a2-03b62b884931",
+		AvailabilityZone: "zone1-groupa",
+		UserData:         []byte("user_data"),
+		RaidArrays: []servers.CreateOptsRaidArray{
```
Comment (t-wata) [inline ecl/baremetal/v2/servers/testing/requests_test.go:87]: `raid_level` parameter is missing.

**Review by t-wata — CHANGES_REQUESTED** [review]
Great Job. Please check my comments.

File: ecl/baremetal/v2/availabilityzones/testing/fixtures.go (hunk 3 above)
Comment (ghost) [reply ecl/baremetal/v2/availabilityzones/testing/fixtures.go]: You're right, it's better to check both. I'll modify it!

File: ecl/baremetal/v2/flavors/requests.go (hunk 6 above)
Comment (ghost) [reply ecl/baremetal/v2/flavors/requests.go:32]: I agree with that, but we need to fix all requests.go which has this kind of function.
I've created new ticket on our backlog:)

File: ecl/baremetal/v2/servers/testing/requests_test.go (hunk 14 above)
Comment (ghost) [reply ecl/baremetal/v2/servers/testing/requests_test.go:87]: got it!

File: ecl/baremetal/v2/servers/requests.go (hunk 8 above)
Comment (ghost) [reply ecl/baremetal/v2/servers/requests.go:53]: commented.

File: ecl/baremetal/v2/servers/testing/requests_test.go (hunk 14 above)
Comment (ghost) [reply ecl/baremetal/v2/servers/testing/requests_test.go:87]: done

**Review by t-wata — APPROVED** [review]
Thank you for responding to all comments! 😄

**Pull request merged**